# Lychrel Numbers
https://projecteuler.net/problem=55

# Problem Statement
If we take 47, reverse and add, 47 + 74 = 121, which is palindromic.

Not all numbers produce palindromes so quickly. For example,

349 + 943 = 1292,

1292 + 2921 = 4213

4213 + 3124 = 7337

That is, 349 took three iterations to arrive at a palindrome.

Although no one has proved it yet, it is thought that some numbers, like 196, never produce a palindrome. A number that never forms a palindrome through the reverse and add process is called a Lychrel number. Due to the theoretical nature of these numbers, and for the purpose of this problem, we shall assume that a number is Lychrel until proven otherwise. In addition you are given that for every number below ten-thousand, it will either (i) become a palindrome in less than fifty iterations, or, (ii) no one, with all the computing power that exists, has managed so far to map it to a palindrome. In fact, 10677 is the first number to be shown to require over fifty iterations before producing a palindrome: 4668731596684224866951378664 (53 iterations, 28-digits).

Surprisingly, there are palindromic numbers that are themselves Lychrel numbers; the first example is 4994.

How many Lychrel numbers are there below ten-thousand?

NOTE: Wording was modified slightly on 24 April 2007 to emphasise the theoretical nature of Lychrel numbers.

# Planning
Write a function that chekcs if a number is a palindrome

Write a function that sums a number with its reverse

Write a function that returns true if a number achieves a palindrome under 50 iterations

In [1]:
def is_palindrome(x):
    """
    x: string
    """
    
    if len(x) <= 1:
        return True
    
    else:
        if x[0] == x[-1]:
            return is_palindrome(x[1:-1])
        else:
            return False
        
        

In [2]:
## testing
is_palindrome('bob')

True

In [3]:
## testing
is_palindrome('carl')

False

In [4]:
def reverse(x):
    """
    x: int
    """
    
    x_str = str(x)
    x_str_reverse = x_str[-1::-1]
    
    return int(x_str_reverse)

In [5]:
reverse(12345)

54321

In [6]:
def reverse_and_sum(x):
    
    return x + reverse(x)

In [7]:
def is_lychrel(x):
    
    reverse_sum = reverse_and_sum(x)
    counts = 1
    
    while not is_palindrome(str(reverse_sum)):
        reverse_sum = reverse_and_sum(reverse_sum)
        
        if counts >= 50:
            return True
        
        counts += 1
        
    return False


In [8]:
## testing
is_lychrel(4994)

True

In [9]:
## testing
is_lychrel(11)

False

In [10]:
# solution

# iterate over all values under 10,000
lychrel_count = 0

for i in range(10000):
    if is_lychrel(i):
        lychrel_count +=1
        
print(lychrel_count)

249


# Solution
There are 249 numbers under 10,000 whose reverse sum sequence does not reach a palindrome in under 50 iterations

Time taken: ~8 to ~8:30